pipelines/mining/cytomining.wdl

```diff
@@ -8,175 +8,280 @@ version 1.0
+            """)
+
+    if not external_metadata:
+        metadata_df = platemap_df
+    else:
+        try:
+            external_df = load_platemap(external_metadata, add_metadata_id=True)
+            metadata_df = platemap_df.merge(
+                external_df,
+                left_on=add_prefix_if_missing(external_join_left),
+                right_on=add_prefix_if_missing(external_join_right),
+                how="left",
+            ).reset_index(drop=True).drop_duplicates()
+        except Exception as err:
+            print(f"Unexpected {err}, {type(err)}")
+            print(f"Unable to merge external metadata with platemap. Check the files and the column names to use for joining them.")
+            raise
+
+    subset_column = add_prefix_if_missing("~{normalize_across_subset_column}")
+    if not subset_column in metadata_df.columns:
+        raise ValueError("""Unable to normalize over a subset of the wells.
+            Metadata contains neither column '~{normalize_across_subset_column}' nor its name with prefix 'Metadata_' added, if not already present.
+            """)
+    if not metadata_df[subset_column].str.contains("~{normalize_across_subset_value}").any():
+        raise ValueError("""Unable to normalize over a subset of the wells.
+            Ensure that metadata column '~{normalize_across_subset_column}', contains at least one '~{normalize_across_subset_value}' value.""")
+
+    metadata_df.to_csv("~{merged_metadata_filename}", index=False)
+    CODE
+
+
+    echo "-----[ Checking for write permissions on output bucket. ]-----"
+    output_url="~{output_directory}"
+    if [[ ${output_url} == "gs://"* ]]; then
+        bucket_name=$(echo "${output_url#gs://}" | sed 's/\/.*//')
+        api_call="https://storage.googleapis.com/storage/v1/b/${bucket_name}/iam/testPermissions?permissions=storage.objects.create"
+        set +o xtrace  # Don't log the bearer access token.
+        bearer=$(gcloud auth application-default print-access-token)
+        curl "${api_call}" --no-progress-meter --header "Authorization: Bearer ${bearer}" --header "Accept: application/json" --compressed > response.json
+        set -o xtrace  # Turn tracing back on.
+        python_json_parsing="import sys, json; print(str('storage.objects.create' in json.load(sys.stdin).get('permissions', ['none'])).lower())"
+        permission=$(cat response.json | python -c "${python_json_parsing}")
+        if [[ ${permission} == false ]]; then
+           echo "The specified output_url ${output_url} cannot be written to."
+           echo "You need storage.objects.create permission on the bucket ${bucket_name}"
+           exit 3
+        fi
+    elif [[ ${output_url} == "s3://"* ]]; then
+        setup_aws_access
+        echo "Check that Terra can write files to this S3 folder." > ~{bucket_write_test_filename}
+        aws s3 cp --acl bucket-owner-full-control ~{bucket_write_test_filename} ~{output_directory}/
```

Comment:
@amygdala I searched for a better way to validate that the S3 bucket was writable without actually writing a file, but I did not come across a generalized way to successfully perform that assertion. Please let me know if you have any other suggestions. This does not `rm` the file because I feel it's best to avoid `rm` commands in workflows.

Reply:
Thanks.. and then I assume the workflow errors and fails if the cp fails?
I will keep an eye out for a more general test, but this seems reasonable.
+1 on being cautious about using `rm`s in workflows.

Reply:
Yes, it fails with an error like the following, which I think is pretty clear:
```
upload failed: ./terra_confirm_bucket_writable.txt to s3://cellpainting-gallery/cpg0000-jump-pilot/deflaux-test/terra_confirm_bucket_writable.txt An error occurred (AccessDenied) when calling the PutObject operation: Access Denied
```